cmd/fossa/test.go

```diff
@@ -0,0 +1,127 @@
+
+	"github.com/tidwall/gjson"
+)
+
+var testLogger = logging.MustGetLogger("test")
+
+const PollRequestDelay = 8000
+
+func confirmBuild(config cliConfig, timing int) error {
+	fossaBaseURL, err := url.Parse(config.endpoint)
+	if err != nil {
+		return errors.New("invalid FOSSA endpoint")
+	}
+
+	reqRef, _ := url.Parse("/api/revisions/" + url.PathEscape(config.getVcsLocator()) + "/build")
+	reqURL := fossaBaseURL.ResolveReference(reqRef).String()
+
+	testLogger.Debugf("Querying <%#v>", reqURL)
+	resp, err := makeAPIRequest("PUT", reqURL, nil, config.apiKey)
+	if err != nil {
+		return err
+	}
+
+	buildData := string(resp)
+	buildStatus := gjson.Get(buildData, "task.status").String()
+
+	testLogger.Debugf("Build status returned: %s", buildStatus)
+
+	switch buildStatus {
+	case "":
+		return errors.New("unable to parse build results")
+	case "FAILED":
+		return errors.New("failed to analyze build #" + gjson.Get(buildData, "id").String() + " <" + gjson.Get(buildData, "error").String() + ">; visit FOSSA or contact support@fossa.io")
+	case "SUCCEEDED":
+		return nil
+	default:
+	}
+
+	if timing >= config.timeout {
+		return errors.New("request series timed out")
+	}
+
+	time.Sleep(time.Duration(PollRequestDelay) * time.Millisecond)
+
+	return confirmBuild(config, timing+PollRequestDelay)
+}
+
+func confirmScan(config cliConfig, timing int) error {
+	fossaBaseURL, err := url.Parse(config.endpoint)
+	if err != nil {
+		return errors.New("invalid FOSSA endpoint")
+	}
+
+	reqRef, _ := url.Parse("/api/revisions/" + url.PathEscape(config.getVcsLocator()))
+	reqURL := fossaBaseURL.ResolveReference(reqRef).String()
+
+	testLogger.Debugf("Querying <%#v>", reqURL)
+	resp, err := makeAPIRequest("GET", reqURL, nil, config.apiKey)
+	if err != nil {
+		return err
+	}
+
+	revisionData := string(resp)
+	if gjson.Get(revisionData, "meta.0.last_scan").String() == "" {
+		// not scanned yet
+		if timing >= config.timeout {
+			return errors.New("request series timed out")
+		}
+		time.Sleep(time.Duration(PollRequestDelay) * time.Millisecond)
+
+		return confirmScan(config, timing+PollRequestDelay)
+	}
+
+	issueCount := 0
+	gjson.Get(revisionData, "issues.#.resolved").ForEach(func(key, value gjson.Result) bool {
+		if value.Bool() == false {
+			issueCount++
+		}
+		return true // keep iterating
+	})
+	if issueCount > 0 {
+		return errors.New(strconv.Itoa(issueCount) + " issues found")
+	}
+
+	return nil
+}
+
+func testCmd(c *cli.Context) {
+	config, err := initialize(c)
+	if err != nil {
+		testLogger.Fatalf("Could not load configuration: %s", err.Error())
+	}
+
+	s := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
+	s.Writer = os.Stderr
+	s.Suffix = " Waiting for analysis job to succeed..."
+	s.Start()
+
+	if err := confirmBuild(config, 0); err != nil {
+		s.Stop()
+		testLogger.Fatalf("Error executing test: %#v", err)
+	}
+
+	s.Suffix = " Waiting for FOSSA scan results..."
+	s.Restart()
+
+	if err := confirmScan(config, 0); err != nil {
```

Comment:
Errors are a terrible way to express a return result. This conflates an actual error occurrence with an expected output. Instead `confirmScan` should return `(int, error)`.

Also, `confirmScan` should _really_ just return the whole issues struct so the `testCmd` can do whatever processing it needs. For example, `testCmd` should have a mode where it outputs the reports as machine-readable JSON so that other processes can use its output.

Reply:
I think we should actually pull out the request functions into a separate file, as we may want to use the output for other reporting functions, and we can leave `test` as just an interface for the user.